### PR TITLE
Rewrite how dag to dataset / dataset alias are stored

### PR DESCRIPTION
**Pull request by Lee-W**
## Why

If someone subclasses a dataset or a dataset alias and makes it unhashable. The dag ref will not be able to be handled.

## What
Save the data needed as `(type, URI or name)` format in dag ref

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by potiuk — COMMENTED** [review]
Cautious approve :)

**Review by potiuk — APPROVED** [review]

**Comment by uranusjr** [comment]
Doesn’t `Dataset` do this automatically? We shouldn’t let an unsanitised URI get into the system in the first place.

https://github.com/apache/airflow/blob/cb7697ff0b9df0a84bb56a6d36f9f184cde36949/airflow/datasets/__init__.py#L248-L251

**Comment by Lee-W** [comment]
> Doesn’t `Dataset` do this automatically? We shouldn’t let an unsanitised URI get into the system in the first place.
> 
> https://github.com/apache/airflow/blob/cb7697ff0b9df0a84bb56a6d36f9f184cde36949/airflow/datasets/__init__.py#L248-L251

Yep, but `new_datasets` are `DatasetModel`s instead of `Dataset`

**Comment by uranusjr** [comment]
I was thinking it’s weird we’re using DatasetModel here… (DatasetManager actually converts them back to Dataset for a user hook too)

I was having problems with this function when adding `name` too. Let me rewrite this entire function a bit.

**Comment by Lee-W** [comment]
> I was thinking it’s weird we’re using DatasetModel here… (DatasetManager actually converts them back to Dataset for a user hook too)
> 
> I was having problems with this function when adding `name` too. Let me rewrite this entire function a bit.

I think the main reason is that we need to check the stored dataset models. Do you want me to continue working on this one? Or wait for your PR? Without it, some edges cases might fail

**Comment by uranusjr** [comment]
Hmm, wait, this still does not make sense even considering we’re using DatasetModel. `d` here is a `Dataset`, so `d.uri` should have already gone through `_sanitize_uri` when `d` is created. Is `converter` somehow not functioning the way it’s assumed to? Is there a bug somewhere else?

It’s probably best to come up with a test case here first to identify the actual issue. I don’t think calling `_sanitize_uri` is the correct thing to do, but it’s difficult to tell what the right thing is without an actual bug.

**Comment by Lee-W** [comment]
> Hmm, wait, this still does not make sense even considering we’re using DatasetModel. `d` here is a `Dataset`, so `d.uri` should have already gone through `_sanitize_uri` when `d` is created. Is `converter` somehow not functioning the way it’s assumed to? Is there a bug somewhere else?
> 
> It’s probably best to come up with a test case here first to identify the actual issue. I don’t think calling `_sanitize_uri` is the correct thing to do, but it’s difficult to tell what the right thing is without an actual bug.

https://github.com/astronomer/astro-sdk/blob/main/python-sdk/example_dags/example_datasets.py

This is the DAG that fails and yep, I think the solution is incorrect and that's why I convert it to draft again

**Comment by Lee-W** [comment]
I think I found the root cause. will push in short

**Comment by Lee-W** [comment]
> Hmm, wait, this still does not make sense even considering we’re using DatasetModel. `d` here is a `Dataset`, so `d.uri` should have already gone through `_sanitize_uri` when `d` is created. Is `converter` somehow not functioning the way it’s assumed to? Is there a bug somewhere else?
> 
> It’s probably best to come up with a test case here first to identify the actual issue. I don’t think calling `_sanitize_uri` is the correct thing to do, but it’s difficult to tell what the right thing is without an actual bug.

This was due to some dataset/dataset alias subclass might not be hashable. I rewrote how the dag ref was handled.

**Comment by uranusjr** [comment]
If this is already passing, maybe we can also try to merge it first and do the related changes in a different PR. Do you think this is ready to be merged?

File: airflow/models/dag.py [hunk 1]
```diff
@@ -2805,63 +2804,70 @@ def bulk_write_to_db(
...
+                stored_dataset_models[stored_dataset_model.uri] = stored_dataset_model
             else:
-                new_datasets.append(dataset)
-        dataset_manager.create_datasets(dataset_models=new_datasets, session=session)
-        stored_datasets.update({dataset.uri: dataset for dataset in new_datasets})
+                new_dataset_models.append(dataset)
+        dataset_manager.create_datasets(dataset_models=new_dataset_models, session=session)
+        stored_dataset_models.update(
+            {dataset_model.uri: dataset_model for dataset_model in new_dataset_models}
+        )
```
Comment (uranusjr) [inline airflow/models/dag.py:2834]:
```suggestion
        stored_dataset_models.update(
            (dataset_model.uri, dataset_model) for dataset_model in new_dataset_models
        )
```

Saves some memory

Comment (uranusjr) [reply airflow/models/dag.py:2834]: Same for the alias call below.

**Review by uranusjr — APPROVED** [review]
One nit (actually two but on the same topic), looks good to me.

**Comment by uranusjr** [comment]
Meh, I’m going to just merge this since I have plans to refactor the entire function anyway. I can change the thing I want when that happens.

**Comment by Lee-W** [comment]
Thanks! @uranusjr I think this fix is needed for 2.10 so I'm going to backport now

**Pull request merged**